### PR TITLE
Adapting adjacent_difference to work with proxy iterators

### PR DESCRIPTION
**Pull request by hkaiser**
- fixing assignment operator for proxy objects returned from segmented iterators
- adding proxy_value trait allowing to extract type returned by a proxy
- flyby: modernize adjacent_difference algorithm, adding missing concept checks

### CODE REVIEW

File: components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp [hunk 1]
```diff
@@ -1230,8 +1234,21 @@ namespace hpx { namespace traits {
...
+    struct is_value_proxy<hpx::segmented::detail::vector_value_proxy<T, Data>>
+    {
+        using type = T;
+    };
+
+    template <typename T, typename Data>
+    struct proxy_value<hpx::segmented::detail::vector_value_proxy<T, Data>>
+    {
+        using type = T;
+    };
```
Comment (msimberg) [inline components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp:1253]: Am I misunderstanding these traits or should this be:
```suggestion
    struct proxy_value<hpx::segmented::detail::local_vector_value_proxy<T, Data>>
    {
        using type = T;
    };

    template <typename T, typename Data>
    struct is_value_proxy<hpx::segmented::detail::vector_value_proxy<T, Data>>
      : std::true_type
    {
    };

    template <typename T, typename Data>
    struct proxy_value<hpx::segmented::detail::vector_value_proxy<T, Data>>
    {
        using type = T;
    };
```
?

Or should `is_value_proxy` simply be `true_type` for `T` and `Data` if and only if `proxy_value<T, Data>::type` exists?

**Review by msimberg — COMMENTED** [review]
One question about the traits. Otherwise this looks good to me. Thanks for looking into this!

File: components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp (hunk 1 above)
Comment (hkaiser) [reply components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp:1253]: `is_value_proxy<T>` checks whether `T` is a type that is used as a proxy. `proxy_value<T>` allows to deduce the type that is returned by a proxy (`is_value_proxy<T>` is true). If `is_value_proxy<T>`  evaluates to false, then `proxy_value<T>` returns `T`.

Comment (msimberg) [reply components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp:1253]: All right, that sounds like what I imagined. But then my first question still stands. Currently `struct proxy_value< hpx::segmented::detail::local_vector_value_proxy<T, Data>>` inherits from `true_type` when I would expect it to have a nested typedef `using type = T`. Likewise, `struct is_value_proxy<hpx::segmented::detail::vector_value_proxy<T, Data>>` has a nested typedef `using type = T` when I would expect it to inherit from `true_type`. In other words, I think there's been some sort mistaken transposition...

Comment (hkaiser) [reply components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp:1253]: I'm confused. The idea is for `is_value_proxy<T>` to be a boolean trait usable in an enable_if (or similar). For this reason it is derived from `true_type` or `false_type` depending on whether `T` is a proxy type or not. Note that `true_type` and `false_type` have an embedded `using type = ...;` referring to themselves.

The type `proxy_value<T>` is not meant to be used as a boolean trait, rather to extract the type returned by the proxy. It 'returns' the type that is provided by the proxy (if `is_value_proxy<T>` would return true). For usability reasons, it defaults to simply 'returning' `T` for non-proxy types (if `is_value_proxy<T>` would return false).

Comment (hkaiser) [reply components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp:1253]: Uhh, I think I see now what you're referring to. Yes - good catch, I messed it up. Will fix.

Comment (hkaiser) [reply components/containers/partitioned_vector/include/hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp]: That's fixed now. Thanks again for catching the issue.

**Comment by hkaiser** [comment]
@msimberg we should go ahead and merge this asap as other PRs are failing because of the problems fixed here.

**Review by msimberg — APPROVED** [review]

**Comment by hkaiser** [comment]
bors merge

**Comment by msimberg** [comment]
> This PR was included in a batch that successfully built, but then failed to merge into master. It will not be retried.
> 
> Additional information:
> 
> ```json
> {"message":"2 of 2 required status checks are expected.","documentation_url":"https://docs.github.com/articles/about-protected-branches"}
> ```

Hrm :/ Those should've been done already (I've only required `cscs/gcc-10` for now). Not quite sure what's happening here. Retrying...

bors merge

**Comment by msimberg** [comment]
> This PR was included in a batch that successfully built, but then failed to merge into master. It will not be retried.
> 
> Additional information:
> 
> ```json
> {"message":"2 of 2 required status checks are expected.","documentation_url":"https://docs.github.com/articles/about-protected-branches"}
> ```

Boo, not quite what I was expecting... Merging manually. Maybe it's trying to tell us something useful, but if it is I don't know what it is. Merging manually...

**Comment by hkaiser** [comment]
@msimberg I think it happens because we protected the master branch and bors first merges things into the branch `staged` and then tries to directly commit to master. We might have to add bors to the admins of this repo (not sure if this is a good idea, however, but worth a try).

**Comment by msimberg** [comment]
That should not be the case anymore (see my failed attempts here: https://github.com/STEllAR-GROUP/hpx/pull/5637/#issuecomment-964971164). It's already merged two PRs...

**Pull request merged**